pkg/credentialprovider/external/external.go

```diff
@@ -0,0 +1,175 @@
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/dockerauth"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/dockerauth/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	dockerauth.AddToScheme(scheme)
+	v1alpha1.AddToScheme(scheme)
+}
+
+// Called from kubelet if alpha feature flag is enabled
+func RegisterExternalProviders(pluginConfigFile string) error {
+	externalProviders, err := NewExternalCredentialProvidersForConfig(pluginConfigFile)
+	if err != nil {
+		return err
+	}
+
+	for _, provider := range externalProviders {
+		credentialprovider.RegisterCredentialProvider(provider.name, provider)
+	}
+
+	return nil
+}
+
+func NewExternalCredentialProvidersForConfig(config string) ([]*externalCredentialProvider, error) {
+	credentialProviderConfig, err := readExternalProviderConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var providers []*externalCredentialProvider
+	for _, provider := range credentialProviderConfig.Providers {
+		providers = append(providers, &externalCredentialProvider{
+			name:          provider.Name,
+			imageMatchers: provider.ImageMatchers,
+			args:          convertArgs(provider.ExtraArgs),
+		})
+	}
+
+	return providers, nil
+}
+
+type externalCredentialProvider struct {
+	name          string
+	imageMatchers []string
+	args          []string
+}
+
+func (e *externalCredentialProvider) Provide(image string) credentialprovider.DockerConfig {
+	cfg := make(credentialprovider.DockerConfig)
+
+	foundMatch := false
+	for _, imageMatcher := range e.imageMatchers {
+		if matched, _ := credentialprovider.URLsMatchStr(imageMatcher, image); matched {
+			foundMatch = true
+			break
+		}
+	}
+
+	if !foundMatch {
+		return cfg
+	}
+
+	response, err := e.request(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return cfg
+	}
+
+	cfg[image] = credentialprovider.DockerConfigEntry{
+		Username: *response.Username,
+		Password: *response.Password,
+	}
+	return cfg
+}
+
+func (e *externalCredentialProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *externalCredentialProvider) IsPlugin() bool {
+	return true
+}
+
+func newDockerAuthEncoder(targetVersion schema.GroupVersion) (runtime.Encoder, error) {
+	mediaType := "application/json"
+	info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("unsupported media type %q", mediaType)
+	}
+	return codecs.EncoderForVersion(info.Serializer, targetVersion), nil
+}
+
+func (e *externalCredentialProvider) request(image string) (*dockerauth.DockerAuthPluginResponse, error) {
+	var stdout, stderr, stdin bytes.Buffer
+
+	args := []string{"get-credentials"}
+	args = append(args, e.args...)
+	cmd := exec.Command(e.name, args...)
```

Comment:
Would it make sense to plumb in context? If this ends up being a remote call we may want to be able to cancel it. Especially if the image/container request is cancelled, times out, ...

Reply:
yup, I can add that in